tailnet/configmaps.go

```diff
@@ -0,0 +1,274 @@
+	sync.Cond
+	netmapDirty  bool
+	derpMapDirty bool
+	filterDirty  bool
+	closing      bool
+	phase        phase
+
+	engine    engineConfigurable
+	static    netmap.NetworkMap
+	peers     map[uuid.UUID]*peerLifecycle
+	addresses []netip.Prefix
+	derpMap   *proto.DERPMap
+	logger    slog.Logger
+}
+
+func newConfigMaps(logger slog.Logger, engine engineConfigurable, nodeID tailcfg.NodeID, nodeKey key.NodePrivate, discoKey key.DiscoPublic, addresses []netip.Prefix) *configMaps {
+	pubKey := nodeKey.Public()
+	c := &configMaps{
+		Cond:   *(sync.NewCond(&sync.Mutex{})),
+		logger: logger,
+		engine: engine,
+		static: netmap.NetworkMap{
+			SelfNode: &tailcfg.Node{
+				ID:       nodeID,
+				Key:      pubKey,
+				DiscoKey: discoKey,
+			},
+			NodeKey:    pubKey,
+			PrivateKey: nodeKey,
+			PacketFilter: []filter.Match{{
+				// Allow any protocol!
+				IPProto: []ipproto.Proto{ipproto.TCP, ipproto.UDP, ipproto.ICMPv4, ipproto.ICMPv6, ipproto.SCTP},
+				// Allow traffic sourced from anywhere.
+				Srcs: []netip.Prefix{
+					netip.PrefixFrom(netip.AddrFrom4([4]byte{}), 0),
+					netip.PrefixFrom(netip.AddrFrom16([16]byte{}), 0),
+				},
+				// Allow traffic to route anywhere.
+				Dsts: []filter.NetPortRange{
+					{
+						Net: netip.PrefixFrom(netip.AddrFrom4([4]byte{}), 0),
+						Ports: filter.PortRange{
+							First: 0,
+							Last:  65535,
+						},
+					},
+					{
+						Net: netip.PrefixFrom(netip.AddrFrom16([16]byte{}), 0),
+						Ports: filter.PortRange{
+							First: 0,
+							Last:  65535,
+						},
+					},
+				},
+				Caps: []filter.CapMatch{},
+			}},
+		},
+		peers:     make(map[uuid.UUID]*peerLifecycle),
+		addresses: addresses,
+	}
+	go c.configLoop()
+	return c
+}
+
+func (c *configMaps) configLoop() {
+	c.L.Lock()
+	defer c.L.Unlock()
+	defer func() {
+		c.phase = closed
+		c.Broadcast()
+	}()
+	for {
+		for !(c.closing || c.netmapDirty || c.filterDirty || c.derpMapDirty) {
+			c.phase = idle
+			c.Wait()
+		}
+		if c.closing {
+			return
+		}
+		// queue up the reconfiguration actions we will take while we have
+		// the configMaps locked. We will execute them while unlocked to avoid
+		// blocking during reconfig.
+		actions := make([]func(), 0, 3)
+		if c.derpMapDirty {
+			derpMap := c.derpMapLocked()
+			actions = append(actions, func() {
+				c.engine.SetDERPMap(derpMap)
+			})
+		}
+		if c.netmapDirty {
+			nm := c.netMapLocked()
+			actions = append(actions, func() {
+				c.engine.SetNetworkMap(nm)
+				c.reconfig(nm)
+			})
+		}
+		if c.filterDirty {
+			f := c.filterLocked()
+			actions = append(actions, func() {
+				c.engine.SetFilter(f)
+			})
+		}
+
+		c.netmapDirty = false
+		c.filterDirty = false
+		c.derpMapDirty = false
+		c.phase = configuring
+		c.Broadcast()
+		func() {
+			// this may look a little odd, but here we want this code to run
+			// without the lock, and then relock when done
+			c.L.Unlock()
+			defer c.L.Lock()
+			for _, a := range actions {
+				a()
+			}
+		}()
```

Comment:
The inline function is a nice way to be able to use defer, but it seems pretty trivial to just put the lock and unlock on either side of the for loop since there's only one entry and exit for the loop. Is there a different reason to needing the defer?

Reply:
Just stylistic.  I can rearrange